File: backend/src/main/java/com/ocean/shopping/service/CouponService.java
package com.ocean.shopping.service;

import com.ocean.shopping.dto.coupon.CouponResponse;
import com.ocean.shopping.dto.coupon.CouponValidationRequest;
import com.ocean.shopping.dto.coupon.CouponValidationResponse;
import com.ocean.shopping.exception.BadRequestException;
import com.ocean.shopping.exception.ResourceNotFoundException;
import com.ocean.shopping.model.entity.*;
import com.ocean.shopping.model.entity.enums.CouponStatus;
import com.ocean.shopping.model.entity.enums.CouponType;
import com.ocean.shopping.repository.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.ZonedDateTime;
import java.util.Optional;

/**
 * Service for coupon operations and discount calculations
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class CouponService {

    private final CouponRepository couponRepository;
    private final OrderCouponRepository orderCouponRepository;
    private final StoreRepository storeRepository;
    private final OrderRepository orderRepository;

    /**
     * Validate a coupon code and calculate discount
     */
    @Transactional(readOnly = true)
    public CouponValidationResponse validateCoupon(CouponValidationRequest request) {
        log.info("Validating coupon code: {}", request.getCode());

        try {
            // Find the coupon
            Optional<Coupon> couponOpt = request.getStoreId() != null 
                ? findCouponForStore(request.getCode(), request.getStoreId())
                : couponRepository.findUsableCouponByCode(request.getCode(), ZonedDateTime.now());

            if (couponOpt.isEmpty()) {
                return CouponValidationResponse.invalid(request.getCode(), "Coupon code not found or not valid");
            }

            Coupon coupon = couponOpt.get();

            // Validate coupon can be used
            String validationError = validateCouponUsage(coupon, request);
            if (validationError != null) {
                return CouponValidationResponse.invalid(request.getCode(), validationError);
            }

            // Calculate discount
            BigDecimal discountAmount = calculateDiscount(coupon, request.getOrderAmount());
            boolean freeShipping = coupon.getType() == CouponType.FREE_SHIPPING;

            return CouponValidationResponse.valid(
                coupon.getCode(),
                coupon.getName(),
                coupon.getType(),
                discountAmount,
                request.getOrderAmount(),
                request.getCurrency(),
                freeShipping
            );

        } catch (Exception e) {
            log.error("Error validating coupon {}: {}", request.getCode(), e.getMessage(), e);
            return CouponValidationResponse.invalid(request.getCode(), "An error occurred while validating the coupon");
        }
    }

    /**
     * Apply coupon to an order
     */
    @Transactional
    public OrderCoupon applyCouponToOrder(Order order, String couponCode) {
        log.info("Applying coupon {} to order {}", couponCode, order.getOrderNumber());

        // Validate the coupon
        CouponValidationRequest validationRequest = new CouponValidationRequest();
        validationRequest.setCode(couponCode);
        validationRequest.setOrderAmount(order.getSubtotal()); // Use subtotal before discount
        validationRequest.setStoreId(order.getStore().getId());
        validationRequest.setCustomerEmail(order.getCustomerEmail());
        validationRequest.setCurrency(order.getCurrency());

        CouponValidationResponse validation = validateCoupon(validationRequest);
        if (!validation.isValid()) {
            throw new BadRequestException(validation.getErrorMessage());
        }

        // Find the coupon entity
        Coupon coupon = couponRepository.findUsableCouponByCodeForStore(
            couponCode, 
            order.getStore(), 
            ZonedDateTime.now()
        ).orElseThrow(() -> new ResourceNotFoundException("Coupon not found: " + couponCode));

        // Check if coupon is already applied to this order
        if (orderCouponRepository.existsByOrderAndCoupon(order, coupon)) {
            throw new BadRequestException("Coupon is already applied to this order");
        }

        // Create OrderCoupon record
        OrderCoupon orderCoupon = OrderCoupon.builder()
            .order(order)
            .coupon(coupon)
            .couponCode(coupon.getCode())
            .couponName(coupon.getName())
            .discountAmount(validation.getDiscountAmount())
            .originalOrderAmount(order.getSubtotal())
            .currency(order.getCurrency())
            .build();

        orderCoupon = orderCouponRepository.save(orderCoupon);

        // Update coupon usage
        coupon.incrementUsage();
        couponRepository.save(coupon);

        // Update order discount amount
        BigDecimal newDiscountAmount = order.getDiscountAmount().add(validation.getDiscountAmount());
        order.setDiscountAmount(newDiscountAmount);
        
        // Recalculate total (subtotal + tax + shipping - discount)
        BigDecimal newTotal = order.getSubtotal()
            .add(order.getTaxAmount())
            .add(order.getShippingAmount())
            .subtract(newDiscountAmount);
        order.setTotalAmount(newTotal);

        orderRepository.save(order);

        log.info("Successfully applied coupon {} to order {} with discount {}", 
                couponCode, order.getOrderNumber(), validation.getDiscountAmount());

        return orderCoupon;
    }

    /**
     * Remove coupon from order
     */
    @Transactional
    public void removeCouponFromOrder(Order order, String couponCode) {
        log.info("Removing coupon {} from order {}", couponCode, order.getOrderNumber());

        // Find the coupon first
        Coupon coupon = couponRepository.findByCode(couponCode)
            .orElseThrow(() -> new ResourceNotFoundException("Coupon not found: " + couponCode));
        
        OrderCoupon orderCoupon = orderCouponRepository.findByOrderAndCoupon(order, coupon)
            .orElseThrow(() -> new ResourceNotFoundException("Coupon not applied to this order: " + couponCode));

        // Update order totals
        BigDecimal newDiscountAmount = order.getDiscountAmount().subtract(orderCoupon.getDiscountAmount());
        order.setDiscountAmount(newDiscountAmount);
        
        BigDecimal newTotal = order.getSubtotal()
            .add(order.getTaxAmount())
            .add(order.getShippingAmount())
            .subtract(newDiscountAmount);
        order.setTotalAmount(newTotal);

        // Decrement coupon usage
        Coupon coupon = orderCoupon.getCoupon();
        coupon.setTimesUsed(Math.max(0, coupon.getTimesUsed() - 1));
        if (coupon.getStatus() == CouponStatus.USED_UP && coupon.canBeUsed()) {
            coupon.setStatus(CouponStatus.ACTIVE);
        }

        // Save changes
        orderRepository.save(order);
        couponRepository.save(coupon);
        orderCouponRepository.delete(orderCoupon);

        log.info("Successfully removed coupon {} from order {}", couponCode, order.getOrderNumber());
    }

    /**
     * Get coupon by code
     */
    @Transactional(readOnly = true)
    public CouponResponse getCouponByCode(String code) {
        Coupon coupon = couponRepository.findByCode(code)
            .orElseThrow(() -> new ResourceNotFoundException("Coupon not found: " + code));
        return convertToResponse(coupon);
    }

    /**
     * Get all active coupons for a store
     */
    @Transactional(readOnly = true)
    public Page<CouponResponse> getActiveCouponsForStore(Long storeId, Pageable pageable) {
        Store store = storeRepository.findById(storeId)
            .orElseThrow(() -> new ResourceNotFoundException("Store not found: " + storeId));
        
        Page<Coupon> coupons = couponRepository.findByStoreAndStatus(store, CouponStatus.ACTIVE, pageable);
        return coupons.map(this::convertToResponse);
    }

    /**
     * Get all global coupons
     */
    @Transactional(readOnly = true)
    public Page<CouponResponse> getGlobalCoupons(Pageable pageable) {
        Page<Coupon> coupons = couponRepository.findByStoreIsNullAndStatus(CouponStatus.ACTIVE, pageable);
        return coupons.map(this::convertToResponse);
    }

    /**
     * Update coupon statuses (cleanup expired and used up coupons)
     */
    @Transactional
    public void updateCouponStatuses() {
        ZonedDateTime now = ZonedDateTime.now();

        // Mark expired coupons
        couponRepository.findExpiredActiveCoupons(now)
            .forEach(coupon -> {
                coupon.setStatus(CouponStatus.EXPIRED);
                couponRepository.save(coupon);
            });

        // Mark used up coupons
        couponRepository.findUsedUpActiveCoupons()
            .forEach(coupon -> {
                coupon.setStatus(CouponStatus.USED_UP);
                couponRepository.save(coupon);
            });
    }

    // Private helper methods

    private Optional<Coupon> findCouponForStore(String code, Long storeId) {
        Store store = storeRepository.findById(storeId).orElse(null);
        if (store == null) {
            return Optional.empty();
        }
        return couponRepository.findUsableCouponByCodeForStore(code, store, ZonedDateTime.now());
    }

    private String validateCouponUsage(Coupon coupon, CouponValidationRequest request) {
        // Check if coupon can be used
        if (!coupon.canBeUsed()) {
            if (coupon.isExpired()) {
                return "Coupon has expired";
            }
            if (coupon.isUsedUp()) {
                return "Coupon usage limit reached";
            }
            return "Coupon is not active";
        }

        // Check minimum order amount
        if (coupon.getMinimumOrderAmount() != null && 
            request.getOrderAmount().compareTo(coupon.getMinimumOrderAmount()) < 0) {
            return "Order amount must be at least " + coupon.getMinimumOrderAmount() + " " + coupon.getCurrency();
        }

        // Check per-customer usage limit
        if (request.getCustomerEmail() != null && 
            coupon.getUsageLimitPerCustomer() != null && 
            coupon.getUsageLimitPerCustomer() > 0) {
            long customerUsage = couponRepository.countUsageByCustomer(coupon, request.getCustomerEmail());
            if (customerUsage >= coupon.getUsageLimitPerCustomer()) {
                return "You have reached the usage limit for this coupon";
            }
        }

        // Check first-time customer requirement
        if (coupon.getFirstTimeCustomerOnly() && request.getCustomerEmail() != null) {
            boolean hasUsedCoupons = orderCouponRepository.hasCustomerUsedAnyCoupon(request.getCustomerEmail());
            if (hasUsedCoupons) {
                return "This coupon is only available for first-time customers";
            }
        }

        return null; // No validation errors
    }

    private BigDecimal calculateDiscount(Coupon coupon, BigDecimal orderAmount) {
        return coupon.calculateDiscount(orderAmount);
    }

    private CouponResponse convertToResponse(Coupon coupon) {
        return CouponResponse.builder()
            .id(coupon.getId())
            .code(coupon.getCode())
            .name(coupon.getName())
            .description(coupon.getDescription())
            .type(coupon.getType())
            .status(coupon.getStatus())
            .storeId(coupon.getStore() != null ? coupon.getStore().getId() : null)
            .storeName(coupon.getStore() != null ? coupon.getStore().getName() : null)
            .discountPercentage(coupon.getDiscountPercentage())
            .discountAmount(coupon.getDiscountAmount())
            .minimumOrderAmount(coupon.getMinimumOrderAmount())
            .maximumDiscount(coupon.getMaximumDiscount())
            .usageLimit(coupon.getUsageLimit())
            .usageLimitPerCustomer(coupon.getUsageLimitPerCustomer())
            .timesUsed(coupon.getTimesUsed())
            .validFrom(coupon.getValidFrom())
            .validUntil(coupon.getValidUntil())
            .currency(coupon.getCurrency())
            .appliesToSaleItems(coupon.getAppliesToSaleItems())
            .firstTimeCustomerOnly(coupon.getFirstTimeCustomerOnly())
            .isActive(coupon.isActive())
            .isExpired(coupon.isExpired())
            .isUsedUp(coupon.isUsedUp())
            .createdAt(coupon.getCreatedAt())
            .updatedAt(coupon.getUpdatedAt())
            .build();
    }
}

File: frontend/src/services/couponService.ts
import api from './api';

// Coupon types and enums
export enum CouponType {
  PERCENTAGE = 'PERCENTAGE',
  FIXED_AMOUNT = 'FIXED_AMOUNT',
  FREE_SHIPPING = 'FREE_SHIPPING',
  BUY_ONE_GET_ONE = 'BUY_ONE_GET_ONE'
}

export enum CouponStatus {
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
  EXPIRED = 'EXPIRED',
  USED_UP = 'USED_UP'
}

// Coupon validation interfaces
export interface CouponValidationRequest {
  code: string;
  orderAmount: number;
  storeId?: number;
  customerEmail?: string;
  currency?: string;
}

export interface CouponValidationResponse {
  valid: boolean;
  code?: string;
  name?: string;
  type?: CouponType;
  discountAmount: number;
  originalAmount?: number;
  finalAmount?: number;
  discountPercentage?: number;
  currency?: string;
  freeShipping?: boolean;
  errorMessage?: string;
  message?: string;
}

// Coupon details interface
export interface Coupon {
  id: number;
  code: string;
  name: string;
  description?: string;
  type: CouponType;
  status: CouponStatus;
  storeId?: number;
  storeName?: string;
  discountPercentage?: number;
  discountAmount?: number;
  minimumOrderAmount?: number;
  maximumDiscount?: number;
  usageLimit?: number;
  usageLimitPerCustomer?: number;
  timesUsed: number;
  validFrom: string;
  validUntil: string;
  currency: string;
  appliesToSaleItems: boolean;
  firstTimeCustomerOnly: boolean;
  isActive: boolean;
  isExpired: boolean;
  isUsedUp: boolean;
  createdAt: string;
  updatedAt: string;
}

// API response interfaces
export interface CouponListResponse {
  content: Coupon[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
  first: boolean;
  last: boolean;
  empty: boolean;
}

class CouponService {
  /**
   * Validate a coupon code and get discount details
   */
  async validateCoupon(request: CouponValidationRequest): Promise<CouponValidationResponse> {
    try {
      const response = await api.post<CouponValidationResponse>('/coupons/validate', request);
      return response.data;
    } catch (error: any) {
      console.error('Error validating coupon:', error);
      return {
        valid: false,
        code: request.code,
        discountAmount: 0,
        errorMessage: error.response?.data?.message || 'Failed to validate coupon'
      };
    }
  }

  /**
   * Quick validation to check if coupon exists and is valid
   */
  async quickValidate(code: string, orderAmount?: number, storeId?: number): Promise<boolean> {
    try {
      const params = new URLSearchParams();
      if (orderAmount !== undefined) params.append('orderAmount', orderAmount.toString());
      if (storeId !== undefined) params.append('storeId', storeId.toString());

      const url = `/coupons/quick-validate/${encodeURIComponent(code)}?${params.toString()}`;
      const response = await api.get<boolean>(url);
      return response.data;
    } catch (error) {
      console.error('Error in quick coupon validation:', error);
      return false;
    }
  }

  /**
   * Get discount preview for a coupon without applying it
   */
  async getDiscountPreview(
    code: string, 
    orderAmount: number, 
    storeId?: number, 
    customerEmail?: string
  ): Promise<CouponValidationResponse> {
    try {
      const params = new URLSearchParams();
      params.append('orderAmount', orderAmount.toString());
      if (storeId !== undefined) params.append('storeId', storeId.toString());
      if (customerEmail) params.append('customerEmail', customerEmail);

      const url = `/coupons/${encodeURIComponent(code)}/preview?${params.toString()}`;
      const response = await api.get<CouponValidationResponse>(url);
      return response.data;
    } catch (error: any) {
      console.error('Error getting discount preview:', error);
      return {
        valid: false,
        code,
        discountAmount: 0,
        errorMessage: error.response?.data?.message || 'Failed to get discount preview'
      };
    }
  }

  /**
   * Get coupon details by code (requires authentication)
   */
  async getCouponByCode(code: string): Promise<Coupon> {
    try {
      const response = await api.get<Coupon>(`/coupons/${encodeURIComponent(code)}`);
      return response.data;
    } catch (error: any) {
      console.error('Error getting coupon details:', error);
      throw new Error(error.response?.data?.message || 'Failed to get coupon details');
    }
  }

  /**
   * Get global coupons (available to all customers)
   */
  async getGlobalCoupons(page: number = 0, size: number = 20): Promise<CouponListResponse> {
    try {
      const params = new URLSearchParams();
      params.append('page', page.toString());
      params.append('size', size.toString());
      params.append('sort', 'createdAt,desc');

      const response = await api.get<CouponListResponse>(`/coupons/global?${params.toString()}`);
      return response.data;
    } catch (error: any) {
      console.error('Error getting global coupons:', error);
      throw new Error(error.response?.data?.message || 'Failed to get global coupons');
    }
  }

  /**
   * Get store-specific coupons (requires authentication)
   */
  async getStoreCoupons(storeId: number, page: number = 0, size: number = 20): Promise<CouponListResponse> {
    try {
      const params = new URLSearchParams();
      params.append('page', page.toString());
      params.append('size', size.toString());
      params.append('sort', 'createdAt,desc');

      const response = await api.get<CouponListResponse>(`/coupons/store/${storeId}?${params.toString()}`);
      return response.data;
    } catch (error: any) {
      console.error('Error getting store coupons:', error);
      throw new Error(error.response?.data?.message || 'Failed to get store coupons');
    }
  }

  /**
   * Check if a coupon is still valid (client-side validation)
   */
  isCouponValid(coupon: Coupon): boolean {
    const now = new Date();
    const validFrom = new Date(coupon.validFrom);
    const validUntil = new Date(coupon.validUntil);
    
    return coupon.status === CouponStatus.ACTIVE &&
           now >= validFrom &&
           now <= validUntil &&
           (coupon.usageLimit === null || coupon.usageLimit === undefined || coupon.timesUsed < coupon.usageLimit);
  }

  /**
   * Format coupon discount for display
   */
  formatDiscount(coupon: Coupon): string {
    switch (coupon.type) {
      case CouponType.PERCENTAGE:
        return `${coupon.discountPercentage}% off`;
      case CouponType.FIXED_AMOUNT:
        return `$${coupon.discountAmount} off`;
      case CouponType.FREE_SHIPPING:
        return 'Free shipping';
      case CouponType.BUY_ONE_GET_ONE:
        return 'Buy one, get one free';
      default:
        return 'Discount applied';
    }
  }

  /**
   * Get human-readable coupon restrictions
   */
  getCouponRestrictions(coupon: Coupon): string[] {
    const restrictions: string[] = [];
    
    if (coupon.minimumOrderAmount && coupon.minimumOrderAmount > 0) {
      restrictions.push(`Minimum order: $${coupon.minimumOrderAmount}`);
    }
    
    if (coupon.usageLimit && coupon.usageLimit > 0) {
      const remaining = coupon.usageLimit - coupon.timesUsed;
      restrictions.push(`${remaining} uses remaining`);
    }
    
    if (coupon.usageLimitPerCustomer && coupon.usageLimitPerCustomer > 0) {
      restrictions.push(`Limit ${coupon.usageLimitPerCustomer} per customer`);
    }
    
    if (coupon.firstTimeCustomerOnly) {
      restrictions.push('First-time customers only');
    }
    
    if (!coupon.appliesToSaleItems) {
      restrictions.push('Excludes sale items');
    }
    
    const validUntil = new Date(coupon.validUntil);
    restrictions.push(`Expires: ${validUntil.toLocaleDateString()}`);
    
    return restrictions;
  }

  /**
   * Calculate savings amount from discount percentage
   */
  calculateSavings(orderAmount: number, discountPercentage: number): number {
    return (orderAmount * discountPercentage) / 100;
  }

  /**
   * Validate coupon code format (client-side)
   */
  isValidCouponFormat(code: string): boolean {
    // Must be 3-50 characters, uppercase letters, numbers, hyphens, underscores
    const regex = /^[A-Z0-9_-]{3,50}$/;
    return regex.test(code);
  }
}

export const couponService = new CouponService();
export default couponService;